src/main/kotlin/app/revanced/patcher/Context.kt

```diff
@@ -0,0 +1,114 @@
+package app.revanced.patcher
+
+import app.revanced.patcher.apk.Apk
+import app.revanced.patcher.apk.ApkBundle
+import app.revanced.patcher.apk.ResourceFile
+import app.revanced.patcher.util.method.MethodWalker
+import org.jf.dexlib2.iface.Method
+import org.w3c.dom.Document
+import java.io.Closeable
+import java.io.InputStream
+import java.io.OutputStream
+import javax.xml.parsers.DocumentBuilderFactory
+import javax.xml.transform.TransformerFactory
+import javax.xml.transform.dom.DOMSource
+import javax.xml.transform.stream.StreamResult
+
+/**
+ * A common class to constrain [Context] to [BytecodeContext] and [ResourceContext].
+ * @param apkBundle the [ApkBundle] for this context.
+ */
+sealed class Context(val apkBundle: ApkBundle)
+
+/**
+ * A context for the bytecode of an [Apk.Base] file.
+ *
+ * @param apkBundle The [ApkBundle] for this context.
+ */
+class BytecodeContext internal constructor(apkBundle: ApkBundle) : Context(apkBundle) {
+    /**
+     * The list of classes.
+     */
+    val classes = apkBundle.base.bytecodeData.classes
+
+    /**
+     * Create a [MethodWalker] instance for the current [BytecodeContext].
+     *
+     * @param startMethod The method to start at.
+     * @return A [MethodWalker] instance.
+     */
+    fun traceMethodCalls(startMethod: Method) = MethodWalker(this, startMethod)
+}
+
+/**
+ * A context for [Apk] file resources.
+ *
+ * @param apkBundle the [ApkBundle] for this context.
+ */
+class ResourceContext internal constructor(apkBundle: ApkBundle) : Context(apkBundle) {
+
+    /**
+     * Open an [DomFileEditor] for a given DOM file.
+     *
+     * @param inputStream The input stream to read the DOM file from.
+     * @return A [DomFileEditor] instance.
+     */
+    fun openXmlFile(inputStream: InputStream) = DomFileEditor(inputStream)
+}
+
+/**
+ * Wrapper for a file that can be edited as a dom document.
+ *
+ * This constructor does not check for locks to the file when writing.
```

Comment:
Looks like I forgot to remove the documentation related to locking after I remove the lock itself.
I removed those docs and simplified the save code because we don't have to do any `OutputStream` workarounds unlike with real files.